File: CPPVersion/AoC2020CPP/AoC2020CPP/Day6.cpp
#include "Day6.h"

Day6::Day6() : DefaultClass(6, true)
{
	std::vector<std::string> inp = FR.ReadLine();
	int count = 0;
	Groups.push_back(std::vector<std::string*>());
	for (int i = 0; i < inp.size(); i++)
	{
		if (inp[i] == "")
		{
			count++;
			Groups.push_back(std::vector<std::string*>());
		}
		else
			Groups[count].push_back(&inp[i]);
	}
	std::cout << "File Reader and format";
	TimeResetClock();
	Part1();
	TimeResetClock();
	Part2();
	TimeClock();
}

void Day6::Part1()
{
	int sum = 0;
	for (int i = 0; i < Groups.size(); i++)
	{
		sum += NumofYes(&Groups[i]);
	}
	std::cout << "Part1: " << sum<<"\n";
}

void Day6::Part2()
{
	int sum = 0;
	for (int i = 0; i < Groups.size(); i++)
	{
		sum += SumofYes(&Groups[i]);
	}
	std::cout << "Part2: " << sum << "\n";

}

int Day6::NumofYes(std::vector<std::string*>* group)
{
	bool Letters[27] = {};
	for (int i = 0; i < group->size(); i++)
	{
		for (int j = 0; j < group->at(i)->size(); j++)
		{
			Letters[(int)group->at(i)->at(j) - 97] = true;
		}
	}
	int sum = 0;
	for (int i = 0; i < 27; i++)
	{
		if(Letters[i])
			sum++;
	}
	return sum;
}

int Day6::SumofYes(std::vector<std::string*>* group)
{
	int Letters[27] = {};
	for (int i = 0; i < group->size(); i++)
	{
		for (int j = 0; j < group->at(i)->size(); j++)
		{
			Letters[(int)group->at(i)->at(j) - 97]++;
		}
	}
	int sum = 0;
	for (int i = 0; i < 27; i++)
	{
		if (Letters[i]==group->size())
			sum++;
	}
	return sum;
}

File: JavaVersion/AoC2020/src/com/company/Day6.java
package com.company;

import java.util.ArrayList;

public class Day6 extends DefaultClass{

    public Day6()
    {
        super(6,true);
        ArrayList<String> inp = fr.ReadLine();
        ArrayList<ArrayList<String>> Groups = new ArrayList<ArrayList<String>>();
        ArrayList<String> temp = new ArrayList<>();
        for (int i = 0; i < inp.size(); i++) {
            if(inp.get(i).equals(""))
            {
                Groups.add(new ArrayList<>(temp));
                temp.clear();
            }
            else
                temp.add(inp.get(i));
        }
        Groups.add(temp);
        System.out.print("File Read and format: ");
        TimeResetClock();
        Part1(Groups);
        TimeResetClock();
        Part2(Groups);
        TimeResetClock();
    }

    private void Part1(ArrayList<ArrayList<String>> Groups)
    {
        int sum = 0;
        for (int i = 0; i < Groups.size(); i++) {
            sum+=NumberOfAnswers(Groups.get(i));
        }
        System.out.println("Part1: "+ sum);
    }

    private int NumberOfAnswers(ArrayList<String> Group)
    {
        boolean letters[] = new boolean[26];
        for (int i = 0; i < Group.size(); i++) {
            for (int j = 0; j < Group.get(i).length(); j++) {
                letters[Group.get(i).charAt(j)-97]=true;
            }
        }
        int sum=0;
        for (int i = 0; i < 26; i++) {
            if(letters[i])
                sum++;
        }
        return sum;
    }

    private void Part2(ArrayList<ArrayList<String>> Groups)
    {
        int sum =0;
        for (int i = 0; i < Groups.size(); i++) {
            sum +=CollectiveAnswer(Groups.get(i));
        }
        System.out.println("Part2: "+ sum);
    }
    private int CollectiveAnswer(ArrayList<String> Group)
    {
        int letters[] = new int[26];
        for (int i = 0; i < Group.size(); i++) {
            for (int j = 0; j < Group.get(i).length(); j++) {
                letters[Group.get(i).charAt(j)-97]++;
            }
        }
        int sum=0;
        for (int i = 0; i < 26; i++) {
            if(letters[i]==Group.size())
                sum++;
        }
        return sum;
    }





}
